README.md
# Py3tftp

Py3tftp is an asynchronous [TFTP][1] server written in Python 3. It was written for the pure joy of working with Python 3 and implements [RFC 1350][2] (except _mail_ mode), [RFC 2347][3] (options), [RFC 2348][4] (blksize option), [RFC 2349][5] (timeout, tsize), and [RFC 7440][10] (windowsize) for RRQ. Additionally, it supports block number roll over, so files of any size can be transferred over.

While a toy project, it does adhere to enough of the standards to be useful in real life.

Some Py3k stuff it uses:
- asyncio - [Transports and Protocols][6] for networking.
- asyncio - [Tasks][7] for spinning up extra handlers.
- [New unpacking methods][8] - some sweet stuff right there (3.5+)
- [Tracebacks attached to exceptions][9] - woo!
- Strings are now bytes because all text is unicode


**This project is in maintenance mode!**

It is not actively developed, but it provides a solid minimum set of functionality and works on Python versions 3.5 though 3.9.

If you would like more functionality or if you find a bug, please open a PR :). 

### Installation

```
pip install py3tftp
```

### Usage

Invoking pyt3tftp will start a server that will interact with the current working directory - it will read and write files from it so don't run it in a place with sensitive files!

TFTP has no security features, except for its simplicity:
- It won't overwrite files.
- Won't create non-existant directories.
- Cannot write outside of the directory it's running from.

```
usage: py3tftp [-h] [--host HOST] [-p PORT] [--ack-timeout ACK_TIMEOUT]
                   [--timeout TIMEOUT] [-l FILE_LOG] [-v] [--version]

optional arguments:
  -h, --help            show this help message and exit
  --host HOST           IP of the interface the server will listen on.
                        Default: 0.0.0.0
  -p PORT, --port PORT  Port the server will listen on. Default: 9069. TFTP
                        standard-compliant port: 69 - requires superuser
                        privileges.
  --ack-timeout ACK_TIMEOUT
                        Timeout for each ACK of the lock-step. Default: 0.5.
  --timeout TIMEOUT     Timeout before the server gives up on a transfer and
                        closes the connection. Default: 3.
  -l FILE_LOG, --file-log FILE_LOG
                        Append output to log file.
  -v, --verbose         Enable debug-level logging.
  --version
```

#### Testing

I wrote some simple acceptance tests in `tests/acceptance/*_test.py`. The code is messy as it's meant to be thrown away.

```
# runs the server with python -m py3tftp
# runs unittests under tests/
# kills the server
./.travis/run.sh
```

#### Extending py3tftp

The way that this module works is that there's a subclass of `BaseTFTPServerProtocol` scheduled on the IO loop that listens for incoming TFTP requests and decides what kind of `BaseTFTPProtocol` to schedule on the IO loop to handle the incoming request.

The way this works in the default scenario is that `TFTPServerProtocol` listens to incoming requests. When a request comes in, it selects either the `WRQProtocol` or the `RRQProtocol` to create a task in the IO loop and passes the request to the selected protocol upon instantiation. From then on, the instantiated protocol handles all of the communication with the client and the _TFTPServerProtocol_ goes back to listening for requests.

This amazing diagram illustrates the above process in the case of a RRQ request:

![py3tftp rrq process diagram](tftp_graph.png)

Extending py3tftp is as easy as:

- Subclassing `BaseTFTPServerProtocol`, mainly to implement the `select_protocol` method to select your custom protocol.
- Subclassing either `RRQProtocol` or `WRQProtocol` to implement your own logic (new options, file handling, etc.) for standard WRQ and RRQ requests, OR...
- Subclassing `BaseTFTPProtocol` to implement your own logic for a custom type of request.

#### Roadplan

(Note, this hasn't been updated in a few years)

- ~~fix off-by-one blksize error ie. if you transfer a file 1000 bytes long and set blksize to 1000 bytes, the server won't ack it.~~
- ~~Pull out file reader/writer from protocol classes~~.
- ~~Add tsize from RFC 2349~~ (added by schrd).
- Add ~~blksize~~, ~~timeout~~, and tsize tests.
- Possibly implement RFCs 906 and 951 for fun!
- Refactor the code, get rid of some duplication, optimize some low-hanging fruit.

#### LICENSE

The MIT License (MIT)

Copyright (c) 2016-2021 sirMackk

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


[1]: https://en.wikipedia.org/wiki/Trivial_File_Transfer_Protocol
[2]: https://tools.ietf.org/html/rfc1350
[3]: https://tools.ietf.org/html/rfc2347
[4]: https://tools.ietf.org/html/rfc2348
[5]: https://tools.ietf.org/html/rfc2349
[6]: https://docs.python.org/3/library/asyncio-protocol.html
[7]: https://docs.python.org/3/library/asyncio-task.html#task
[8]: https://www.python.org/dev/peps/pep-0448/
[9]: http://legacy.python.org/dev/peps/pep-3109/
[10]: https://tools.ietf.org/html/rfc7440


CHANGELOG.md
# CHANGELOG

### 1.3.0 (May 18, 2021)

- Fix timeout interval handling (many thanks pkropine) - https://github.com/sirMackk/py3tftp/pull/17

### 1.2.2 (March 08, 2021)

- Improve filename sanitize to work on Windows (many thanks pkorpine) - https://github.com/sirMackk/py3tftp/pull/15

### 1.2.1 (September 29, 2018)

- Change logger to module logger (thanks schrd!)
- Add flake8 to tests; Fix style issues; Make travis fail on test errors.

### 1.2.0 (September 08, 2018)

- Adds windowsize option handling for RRQ (many thanks jpmzometa!)
- Adds tsize option for RRQ (thanks schrd!)
- handle error packets if client aborts RRQ transfer (thanks schrd!)

### 1.1.0 (August 24, 2018)

- Fixes RRQ not waiting for ACK of last package (thanks jpmzometa!)
- Adds netascii support (thanks keenser!)

### 1.0.2 (March 05, 2018)

- Fixes KeyError bug with the `-l` CLI option (thanks pritstift!)

### 1.0.1 (February 10, 2018)

- Fixes issue #3 - --ack-timeout and --conn-timeout actually work now.

### 1.0.0a (November 26, 2017)

- Fixes issue #1 - implements block number roll over so files over 65535 blocks can be uploaded or downloaded.
- Version bump to 1.0.0a due to fouled up pypi upload.

### 0.0.6 (October 17th, 2016)

- Removes type hints.
- Restructures project into more layers (ie. special packet objects, pulls out file IO from protocols), big API changes.
- Adds a suite of unit tests.
- Adds Travis-CI configuration files.

### 0.0.5 (March 21st, 2016)

- Fixes off by one server response to RRQs that are the multiple of the file's size.
- Adds type hints.
- Makes 'long description' rst.
- Adds test.sh to make running tests easier.

### 0.0.4 (March x, 2016)

- Adds CHANGELOG.md
- Fixes blksize errors.
- Adds tests for blksize and timeout options.
- Improves tftp option handling.


### 0.0.3 (March 1st, 2016)

- Restructuring for easier reuse.
- Fixed leftover file descriptor issue.

###0.0.2 (February 25th, 2016)

- Adds rfc2347 support.
- Adds rfc2348 (blksize) support.
- Adds part of rfc2349 support (timeout, no tsize).
- Adds final setup.py stuff.


py3tftp/cli_parser.py
import argparse
import logging
from sys import exit

from py3tftp import __version__

EPILOG = """
Released under the MIT license.
Copyright 2016 Matt O. <matt@mattscodecave.com>
"""

logging_config = {
    'format': '%(asctime)s [%(levelname)s] %(message)s',
    'level': logging.INFO,
    'filename': None
}


def print_version():
    print("py3tftp version: {}".format(__version__))


def parse_cli_arguments():
    parser = argparse.ArgumentParser(epilog=EPILOG)
    parser.add_argument('--host',
                        default='0.0.0.0',
                        help=('IP of the interface the server will listen on. '
                              'Default: 0.0.0.0'))
    parser.add_argument(
        '-p',
        '--port',
        default=9069,
        type=int,
        help=('Port the server will listen on. '
              'Default: 9069. TFTP standard-compliant port: 69 - '
              'requires superuser privileges.'))
    parser.add_argument(
        '--ack-timeout',
        dest="timeout",
        default=0.5,
        type=float,
        help='Timeout for each ACK of the lock-step. Default: 0.5.')
    parser.add_argument(
        '--conn-timeout',
        default=3.0,
        type=float,
        help=('Timeout before the server gives up on a transfer and closes '
              'the connection. Default: 3.'))
    parser.add_argument('-l', '--file-log', help='Append output to log file.')
    parser.add_argument('-v',
                        '--verbose',
                        action='store_true',
                        help='Enable debug-level logging.')
    parser.add_argument('--version', action='store_true')

    args = parser.parse_args()

    if args.verbose:
        logging_config['level'] = logging.DEBUG

    if args.file_log:
        logging_config['filename'] = args.file_log

    if args.version:
        print_version()
        exit()

    logging.basicConfig(**logging_config)
    return args
